Add lock icon on HTTPS browsing warning page

sysrqb opened the pull request:
#### be3c0c9a661fa4990b468ee62eb4784390a24ebc
<pre>
Add lock icon on HTTPS browsing warning page
<a href="https://bugs.webkit.org/show_bug.cgi?id=280732">https://bugs.webkit.org/show_bug.cgi?id=280732</a>
<a href="https://rdar.apple.com/133800121">rdar://133800121</a>

Reviewed by Aditya Keerthi.

Use the SF Symbols lock.slash icon on the HTTPSOnly browsing warning page,
instead of reusing the exclamation point from the SafeBrowsing page. This
change also replaces the existing exclamation mark with the SF symbol.

This change also has a small drive-by fix for the colors used on non-macOS.

* Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
(colorForItem):
(viewForIconImage):
(-[_WKWarningView addContent]):
(-[WKSafeBrowsingExclamationPoint drawRect:]): Deleted.
(-[WKSafeBrowsingExclamationPoint viewDidChangeEffectiveAppearance]): Deleted.
(-[WKSafeBrowsingExclamationPoint intrinsicContentSize]): Deleted.

Canonical link: <a href="https://commits.webkit.org/285045@main">https://commits.webkit.org/285045@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07c713b755bc8c639745a920282fc368cc852de2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71281 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50694 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24054 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75389 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22486 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58494 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22305 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56345 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14806 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74347 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46050 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61424 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36781 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42723 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20827 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64616 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19249 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77111 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15515 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18444 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64058 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15558 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61458 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64044 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12187 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5813 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10938 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46494 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1273 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47565 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48848 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47307 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
```diff
@@ -180,71 +181,180 @@
+#endif
+}
+
+static void addLockIconPath(BezierPath *path)
```
Can we use an SF symbol (and `UIImageView` / `NSImageView`)? Based on the description it seems like this would be `lock.slash.fill`.

sysrqb replied:
Thanks! `lock.slash.fill` was exactly what I wanted. I'm also migrating the exclamation mark to use the SF symbol.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
```diff
@@ -159,8 +161,7 @@
 #else
     UIColor *red = [UIColor colorWithRed:0.998 green:0.239 blue:0.233 alpha:1.0];
     UIColor *darkGray = [UIColor colorWithRed:0.118 green:0.118 blue:0.118 alpha:1.0];
-    UIColor *lighterGray = [UIColor colorWithRed:0.220 green:0.224 blue:0.231 alpha:1.0];
-    UIColor *mediumGray = [UIColor colorWithRed:0.729 green:0.733 blue:0.741 alpha:1.0];
+    UIColor *lighterGray = [UIColor colorWithRed:0.937 green:0.937 blue:0.937 alpha:1.0];
     UIColor *white = [UIColor whiteColor];
```
Seeing hardcoded colors is always concerning 😅

Does this look okay in dark mode? And high contrast mode?

Note that `UIColor` has `+darkGrayColor`, `+lightGrayColor` and `+grayColor`.

sysrqb replied:
Yeah... `+lightGrayColor` is still too dark (70% grayscale, whereas I need something like ~90%). `+systemColor6Color` is basically what I want in light mode, except now I need to adjust the text font color, as well, so it adapts to dark mode. Currently, these warning pages in Safari don't adapt to appearance changes on iOS at all. I'd almost rather not adapt so this View remains consistent with the other warning pages.

pxlcoder replied:
Makes sense, dark mode sounds like a possible future enhancement.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
```diff
@@ -179,88 +180,58 @@
-    ViewType *warning = self.superview.superview;
+    NSString *symbolName;
+    WebCore::CocoaColor *color = colorForItem(WarningItem::WarningSymbol, warningView);
+    Boolean shouldSetTint = NO;
```
```suggestion
    BOOL shouldSetTint = NO;
```

sysrqb replied:
Whoops, I spent too much time in CF code before writing this.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
```diff
@@ -179,88 +180,58 @@
+    NSImageView *view = [NSImageView imageViewWithImage:[NSImage imageWithSystemSymbolName:symbolName accessibilityDescription:nil]];
+    view.symbolConfiguration = [NSImageSymbolConfiguration configurationWithPointSize:imagePointSize weight:NSFontWeightRegular scale:NSImageSymbolScaleLarge];
+    if (shouldSetTint)
+        view.contentTintColor = color;
```
```suggestion
    RetainPtr view = [NSImageView imageViewWithImage:[NSImage imageWithSystemSymbolName:symbolName accessibilityDescription:nil]];
    [view setSymbolConfiguration:[NSImageSymbolConfiguration configurationWithPointSize:imagePointSize weight:NSFontWeightRegular scale:NSImageSymbolScaleLarge]];
    if (shouldSetTint)
        [view setContentTintColor:color];
```

(the current code is safe, but we are trying to use smart pointers where possible)

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
```diff
@@ -179,88 +180,58 @@
-{
-    [self setNeedsDisplay:YES];
-}
+    UIImageView *view = [[UIImageView alloc] initWithImage:[UIImage systemImageNamed:symbolName]];
```
This is a memory leak – need to use `RetainPtr` + `adoptNS`.

pxlcoder reviewed (commented):
This is looking much better 🙂

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
```diff
@@ -179,88 +180,58 @@
-- (NSSize)intrinsicContentSize
-{
-    return { exclamationPointSize, exclamationPointSize };
+    return view.get();
```
```suggestion
    return view;
```

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
```diff
@@ -179,88 +180,58 @@
-@implementation WKSafeBrowsingExclamationPoint
-
-- (void)drawRect:(RectType)rect
+static ViewType* viewForIconImage(_WKWarningView *warningView)
```
```suggestion
static RetainPtr<ViewType> viewForIconImage(_WKWarningView *warningView)
```

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
```diff
@@ -370,7 +341,7 @@ - (instancetype)initWithFrame:(RectType)frame browsingWarning:(const WebKit::Bro
 - (void)addContent
 {
-    auto exclamationPoint = adoptNS([WKSafeBrowsingExclamationPoint new]);
+    ViewType *warningViewIcon = viewForIconImage(self);
```
```suggestion
    RetainPtr warningViewIcon = viewForIconImage(self);
```

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
```diff
@@ -397,7 +368,7 @@ - (void)addContent
     [box layer].cornerRadius = boxCornerRadius;
 
-    for (ViewType *view in @[exclamationPoint.get(), title.get(), warning.get(), goBack, primaryButton]) {
+    for (ViewType *view in @[warningViewIcon, title.get(), warning.get(), goBack, primaryButton]) {
```
```suggestion
    for (ViewType *view in @[ warningViewIcon.get(), title.get(), warning.get(), goBack, primaryButton ]) {
```

pxlcoder reviewed (approved):
Good to land after fixing the smart pointer issues.

Commit:
Add lock icon on HTTPS browsing warning page
https://bugs.webkit.org/show_bug.cgi?id=280732
rdar://133800121

Reviewed by Aditya Keerthi.

Use the SF Symbols lock.slash icon on the HTTPSOnly browsing warning page,
instead of reusing the exclamation point from the SafeBrowsing page. This
change also replaces the existing exclamation mark with the SF symbol.

This change also has a small drive-by fix for the colors used on non-macOS.

* Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
(colorForItem):
(viewForIconImage):
(-[_WKWarningView addContent]):
(-[WKSafeBrowsingExclamationPoint drawRect:]): Deleted.
(-[WKSafeBrowsingExclamationPoint viewDidChangeEffectiveAppearance]): Deleted.
(-[WKSafeBrowsingExclamationPoint intrinsicContentSize]): Deleted.

Canonical link: https://commits.webkit.org/285045@main
